rust/yarp-sys/build.rs

```diff
@@ -0,0 +1,180 @@
+    println!(
+        "cargo:rustc-link-search=native={}",
+        ruby_build_path.display()
+    );
+
+    // This is where the magic happens.
+    let bindings = generate_bindings(&ruby_include_path);
+
+    // Write the bindings to file.
+    write_bindngs(bindings);
+
+    Ok(())
+}
+
+/// Convenience step for running `rake compile` before doing our stuff.
+///
+fn rake_compile() -> Result<(), io::Error> {
+    let output = Command::new("rake").arg("compile").output()?;
+
+    io::stdout().write_all(&output.stdout)?;
+    io::stderr().write_all(&output.stderr)?;
+
+    println!("`rake compile` exit status: {}", output.status);
+
+    Ok(())
+}
+
+/// Gets the path to project files (`librubyparser*`) at `[root]/build/`.
+///
+fn ruby_build_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../build/")
+        .canonicalize()
+        .unwrap()
+}
+
+/// Gets the path to the header files that `bindgen` needs for doing code generation.
+///
+fn ruby_include_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../include/")
+        .canonicalize()
+        .unwrap()
+}
+
+fn cargo_manifest_path() -> &'static Path {
+    Path::new(env!("CARGO_MANIFEST_DIR"))
+}
+
+/// Uses `bindgen` to generate bindings to the C API. Update this to allow new types/functions/etc
+/// to be generated (it's allowlisted to only expose functions that'd make sense for public
+/// consumption).
+///
+/// This method only generates code in memory here--it doesn't write it to file.
+///
+fn generate_bindings(ruby_include_path: &Path) -> bindgen::Bindings {
+    bindgen::Builder::default()
+        .header(ruby_include_path.join("yarp/defines.h").to_string_lossy())
+        .header(ruby_include_path.join("yarp.h").to_string_lossy())
+        .clang_arg(format!("-I{}", ruby_include_path.display()))
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        // Structs
+        .allowlist_type("yp_buffer_t") // TEST ✅
+        .allowlist_type("yp_comment_t") // TEST ✅
+        .allowlist_type("yp_diagnostic_t") // TEST ✅
+        .allowlist_type("yp_encoding_changed_callback_t") // TEST ✅
+        .allowlist_type("yp_encoding_decode_callback_t") // TEST ✅
+        .allowlist_type("yp_memsize_t") // TEST ✅
+        .allowlist_type("yp_list_t") // TEST ✅
+        .allowlist_type("yp_node_t") // TEST ✅
+        .allowlist_type("yp_parser_t") // TEST ✅
+        .allowlist_type("yp_pack_size") // TEST ✅
+        // TODO: Commenting this because I can't figure out how to get bindgen to generate the
+        // inner-unions nicely. Hand-rolling this in src/lib.rs for now.
+        // .allowlist_type("yp_string_t")
+        // TODO bindgen is still generating the inner unions, despite this...
+        .blocklist_type("^yp_string_t*") // TEST ✅
+        .allowlist_type("yp_string_list_t") // TEST ✅
+        .allowlist_type("yp_token_type_t")
+        // Enums
+        .rustified_non_exhaustive_enum("yp_comment_type_t") // TEST ✅
```

Comment:
I don't know if I feel strongly about this choice for `enum`s here, but for my part, I find it more ergonomic to deal with Rust-style enums when working in Rust (as opposed to constants that have integer values), hence the choice for `rustified_`.

`bindgen` info on enum options: https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#enums

---

I went with the `non_exhaustive` enum here because a) I don't know what's planned for `yarp` with regards to these types, and b) `yarp` is still relatively young, so maybe these C enums will gain variants over time? By going with non-exhaustive Rust enums, this forces downstream crates to handle variants that might get added in the future--effectively ensuring forward compatibility if more enum variants are added (adding variants to a standard/exhaustive Rust enum is a breaking change since consumers' code won't handle a new variant and thus won't compile). If we know that the C enums are set in stone, then it'd be nice to change this codegen code to generate exhaustive Rust enums (since those are nicer to deal with in Rust).

tl;dr Going this route takes away a tad bit of ergonomics for consumers, but ensures forward compatibility.

More info on non-exhaustive enums: https://docs.rs/bindgen/latest/bindgen/struct.Builder.html#enums

Reply:
Since Ruby syntax is a moving target, I think non-exhaustive default makes sense for most things.